awswrangler/distributed/datasources/parquet_datasource.py

```diff
@@ -62,109 +45,9 @@ class ParquetDatasource:
-                    columns=columns,
-                    schema=schema,
-                    batch_size=PARQUET_READER_ROW_BATCH_SIZE,
-                )
-                for batch in batches:
-                    # Table creation is wrapped inside _add_table_partitions
-                    # to add columns with partition values when dataset=True
-                    table = _add_table_partitions(
-                        table=pa.Table.from_batches([batch], schema=schema),
-                        path=f"s3://{piece.path}",
-                        path_root=path_root,
-                    )
-                    # If the table is empty, drop it.
-                    if table.num_rows > 0:
-                        output_buffer.add_block(table)
-                        if output_buffer.has_next():
-                            yield output_buffer.next()
-
-            output_buffer.finalize()
-            if output_buffer.has_next():
-                yield output_buffer.next()
-
-        if _block_udf is not None:
-            # Try to infer dataset schema by passing dummy table through UDF.
-            dummy_table = schema.empty_table()
-            try:
-                inferred_schema = _block_udf(dummy_table).schema
-                inferred_schema = inferred_schema.with_metadata(schema.metadata)
-            except Exception:  # pylint: disable=broad-except
-                _logger.debug(
-                    "Failed to infer schema of dataset by passing dummy table "
-                    "through UDF due to the following exception:",
-                    exc_info=True,
-                )
-                inferred_schema = schema
-        else:
-            inferred_schema = schema
-        read_tasks = []
-        metadata = meta_provider.prefetch_file_metadata(parquet_dataset.pieces) or []
-        try:
-            _register_parquet_file_fragment_serialization()  # type: ignore
-            for pieces, metadata in zip(  # type: ignore
-                np.array_split(parquet_dataset.pieces, parallelism),
-                np.array_split(metadata, parallelism),
-            ):
-                if len(pieces) <= 0:
-                    continue
-                serialized_pieces = cloudpickle.dumps(pieces)  # type: ignore
-                input_files = [p.path for p in pieces]
-                meta = meta_provider(
-                    input_files,
-                    inferred_schema,
-                    pieces=pieces,
-                    prefetched_metadata=metadata,
-                )
-                read_tasks.append(ReadTask(lambda p=serialized_pieces: read_pieces(p), meta))  # type: ignore
-        finally:
-            _deregister_parquet_file_fragment_serialization()  # type: ignore
-
-        return read_tasks
+    def create_reader(self, **kwargs: Dict[str, Any]) -> Reader[Any]:
+        """Return a Reader for the given read arguments."""
+        return _ParquetDatasourceReader(**kwargs)  # type: ignore
```

Comment:
Does this reader automatically handle partitioning the way we need it to be handled?

Reply:
Yep, it adds partitions as columns

Reply:
I wish it was this simple but I highly doubt that this class is doing everything that the previous implementation was doing in terms of partitioning:
1. Our method ([_add_table_partitions](https://github.com/aws/aws-sdk-pandas/blob/release-3.0.0/awswrangler/_arrow.py#L36)) not only adds partitions, but also converts them to the categorical type (with `.dictionary_encode()`) like in the non-distributed version
2. I don't see how this would honour the `dataset` equals `True` vs `False` case? You have removed the `path_root` argument from the call which we were [using](https://github.com/aws/aws-sdk-pandas/blob/release-3.0.0/awswrangler/_arrow.py#L33) to distinguish between the two cases. The Ray implementation does not read partitions the way we want it, you can use this script to see the differences:

```
import awswrangler as wr

bucket = "my-bucket"
df = pd.DataFrame({"c0": [0, 1, 2], "c1": [3, 4, 5], "c2": [6, 7, 8]})

wr.s3.delete_objects(f"s3://{bucket}/pq2/")

wr.s3.to_parquet(df=df, path=f"s3://{bucket}/pq2/", dataset=True, partition_cols=["c1", "c2"])

print(wr.s3.read_parquet(path=f"s3://{bucket}/pq2/", dataset=True))

print(wr.s3.read_parquet(path=f"s3://{bucket}/pq2/c1=3/", dataset=True))

print(wr.s3.read_parquet(path=f"s3://{bucket}/pq2/"))

print(wr.s3.read_parquet(path=f"s3://{bucket}/pq2/c1=3/"))
```

If you test the above against your changes vs the current implementation, you will see how they differ in behaviour

Side note, I think this is where not having the parquet tests in the distributed case is causing us issues as the failing tests would have highlighted the above

Reply:
I have actually ran the two.

Current implementation:
```
   c0 c1 c2
0   0  3  6
1   1  4  7
2   2  5  8
   c0 c1 c2
0   0  3  6
   c0
0   0
1   1
2   2
   c0
0   0
```
Partition columns are of type categorical

```
   c0  c1  c2
0   0   3   6
1   1   4   7
2   2   5   8

   c0    c1    c2
0   0  <NA>  <NA>

Error:  File "pyarrow/error.pxi", line 99, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Invalid column index to add field. is thrown for #3

   c0
0   0
```
Partition columns are of type `Int64`

Reply:
Thanks, both good catches. 

(1) - I'll add conversion to categorical which is missing in 2.0 implementation.
(2) - so as I understand the expected behaviour is if `dataset=False` even if there are partitions detected, we should still only load data under that specific prefix and *not* load partitions as columns bc current implementation does greedy partition loading. 

Both doable. Yeah, in the absence of tests covering those scenarios the 2.0 upgrade was looking deceptively easy :)

Reply:
Yes, when `dataset=False`, the partition columns should not be added.

But also notice the difference in behaviour for the second case of `dataset=True`:
```
print(wr.s3.read_parquet(path=f"s3://{bucket}/pq2/c1=3/", dataset=True))
```
i.e. reading from one specific partition.

We expect `c1` and `c2` to be added as partition columns, but for some reason the ray implementation just returns NA